### DISCUSSION TITLE
Geocentric altitude/az of the sun and moon.

### DISCUSSION THREAD

**Discussion by hidp123** [cosinekitty/astronomy#299] [Q&A] [answered]
Normally, alt & az are topocentric. However, how can I get geocentric alt & az. Altitude at least, if not azimuth.

**Comment by cosinekitty** [comment]
This is an interesting question! I know an easy way to calculate the geocentric altitude, but it's not clear to me how to do the azimuth.

My approach would be to call [ObserverVector](https://github.com/cosinekitty/astronomy/tree/master/source/js#ObserverVector) to obtain a vector from the center of the Earth to the observer. You should pass `false` for `ofdate`, to select J2000 coordinates.

Then call [GeoVector](https://github.com/cosinekitty/astronomy/tree/master/source/js#GeoVector) to obtain a vector from the center of the Earth to the celestial body in question.

Now you have two geocentric vectors, one for the observer and one for the celestial body, both in consistent J2000 equatorial coordinates.

Subtract the observer vector from the body vector to obtain a vector from the observer to the body. Now pass the geocentric observer vector and the observer-centric body vector to the [AngleBetween](https://github.com/cosinekitty/astronomy/tree/master/source/js#anglebetweena-b--number) function. This will return an angle in degrees between the two vectors, which is the body's angle with respect to the geocentric zenith. Subtract that angle from 90 degrees to obtain the geocentric altitude angle.

I hope this helps!  By the way, I am curious: why do you want geocentric angles rather than topocentric angles? That is not something I expected anyone would ever want.

**Reply by hidp123** [reply]
Thanks for the reply.
The reason I ask is because Yallop in his paper for the visibility of the crescent mentions the following:

<img src="https://user-images.githubusercontent.com/84683703/230511655-1bf13cbc-63bc-4f47-abe0-0bf65583c543.png" width="75%">

First I thought this was simply the topocentric altitude of the moon without refraction, but the results weren't matching to what Yallop provided in the tables in his paper. Then it was suggested to try geocentric altitude, and taking those results from Swisseph it seemed to match better with what was in the paper.

**Reply by hidp123** [reply]
Hi, I did the following but not sure if it's correct:

```
observer = astronomy.Observer(39, -76.8)
besttime = astronomy.Time.Make(1990, 11, 19, 22, 29, 11)
ObserverVector = astronomy.ObserverVector(besttime, observer, False)
GeoVector = astronomy.GeoVector(astronomy.Body.Moon, besttime, False)
subtraction = ObserverVector - GeoVector
angle = astronomy.AngleBetween(GeoVector, subtraction)
GeoAlt = 90 - angle

print(angle) #179.10724071159515
print(GeoAlt) #-89.10724071159515
```
I am expecting the geocentric altitude to be around 7°.

**Reply by cosinekitty** [reply]
Try changing the following lines:

```python
subtraction = GeoVector - ObserverVector
angle = astronomy.AngleBetween(ObserverVector, subtraction)
```

**Reply by hidp123** [reply]
I got the following result: 6.397123700518691, however, this is similar to the topo altitude of 6.2713496.
On Swiss Ephemeris I got 7.1640678 for geo altitude. Not sure where the difference is.

**Reply by cosinekitty** [reply]
I agree, that is a large discrepancy: 0.77 degrees. Is it possible the Swiss Ephemeris defines the geocentric altitude as the altitude angle as seen from the center of the Earth? For example:

```
>>> 90 - astronomy.AngleBetween(GeoVector, ObserverVector)
7.28988298892304
```

This is still not quite right. So I'm left thinking Swiss Ephemeris's definition of *geocentric altitude* could mean a few different things. Perhaps it is still a geodetic altitude, only from the center of the Earth, not the surface (i.e. with oblateness correction). If that is the case, I need to think a little more how to calculate that. Let me try that and post back here a little later.

**Reply by cosinekitty** [reply]
I tried the following to get the altitude angle of the Moon using the observer's horizontal orientation, but from the point of view of the Earth's center.

```python
import astronomy
observer = astronomy.Observer(39, -76.8)
besttime = astronomy.Time.Make(1990, 11, 19, 22, 29, 11)
geomoon = astronomy.GeoVector(astronomy.Body.Moon, besttime, False)
# This part is an undocumented internal function to convert a vector to (ra, dec) angles...
eq = astronomy._vector2radec([geomoon.x, geomoon.y, geomoon.z], geomoon.t)
hor = astronomy.Horizon(besttime, observer, eq.ra, eq.dec, astronomy.Refraction.Airless)
print(hor.altitude)
```

The result is: 7.239123678611705.

Is it possible to find out what exactly Swiss Ephemeris is doing?

**Reply by hidp123** [reply]
Hi, Swiss Ephemeris docs are here:
https://www.astro.com/ftp/swisseph/doc/swephprg.htm

The functions we are dealing with are [`swe_calc_ut()`](https://www.astro.com/ftp/swisseph/doc/swephprg.htm#_Toc112948958) and [`swe_azalt()`](https://www.astro.com/ftp/swisseph/doc/swephprg.htm#_Toc112948998).
First `swe_calc_ut()` is called to obtain the ecliptic coordinates, then those coordinates are passed on to `swe_azalt()` to obtain horizontal coordinates (az/alt).

Now, the first function, `swe_calc_ut()`, will either return geocentric values (default), or topocentric if the SEFLG_TOPOCTR  iflag is used.

If I use the `SEFLG_TOPOCTR` iflag and pass on the resulting coordinates to `swe_azalt()` the moon topocentric non-refracted altitude is 6.27134955759181. (Astronomy-engine altitude in this case is 6.270148219416555, minor difference, acceptable.)

And, if I leave it default (i.e. I don't put in `SEFLG_TOPOCTR` iflag into `swe_calc_ut()`) and pass on the resulting coordinates to `swe_azalt()` then the moon altitude is 7.16406779367621.

Note: I am using JPL horizon mode (`SEFLG_JPLHOR` iflag) in Swiss Ephemeris. Also, I'm using the [C#](https://github.com/ygrenier/SwissEphNet) port of Swiss Ephemeris.
The official Swiss Ephemeris page is [here](https://www.astro.com), the github page is [here](https://github.com/aloistr/swisseph), and community groups page is [here](https://groups.io/g/swisseph).

**Reply by cosinekitty** [reply]
OK, I will study this over the next couple of days and see if I can help you do the same thing in Astronomy Engine. Are you pretty sure that the value 7.16406779367621 (or something very close to it) is correct for your use case?

**Reply by hidp123** [reply]
It seems to be, as I posted in the screenshot above, the tables given by [Yallop](http://www.astronomycenter.net/pdf/yallop_1997.pdf) are only to 1 decimal place, so hard to gauge precision.
Also, JPL Horizons does not give geocentric values for az/alt, so hard to compare for verification as well.

I saw this: https://github.com/cosinekitty/astronomy/issues/53#issuecomment-562146817, and this: https://github.com/cosinekitty/astronomy/issues/54, not sure if it is connected to what we are doing.

**Reply by cosinekitty** [reply]
#53 and #54 are about work I did adding coordinate transforms using rotation matrices. Basically any vector in any of the supported orientation systems can be converted to a vector in any other orientation system. This could very well be of help in your case; I'm not sure yet, but certainly possible. The documentation section on [coordinate transforms](https://github.com/cosinekitty/astronomy/tree/master/source/python#coordinate-transforms) gives more details if you are interested.

**Reply by cosinekitty** [reply, marked as answer]
OK, I think I figured it out. I was calculating the vectors in J2000 coordinates (EQJ in Astronomy Engine nomenclature), but to get accurate horizontal coordinates, I need to convert them to of-date coordinates (EQD).

```python3
#!/usr/bin/env python3
import astronomy
observer = astronomy.Observer(39, -76.8)
besttime = astronomy.Time.Make(1990, 11, 19, 22, 29, 11)
geomoon = astronomy.GeoVector(astronomy.Body.Moon, besttime, True)
geosun = astronomy.GeoVector(astronomy.Body.Sun, besttime, True)
rot = astronomy.Rotation_EQJ_EQD(besttime)
rotmoon = astronomy.RotateVector(rot, geomoon)
rotsun  = astronomy.RotateVector(rot, geosun)
meq = astronomy.EquatorFromVector(rotmoon)
seq = astronomy.EquatorFromVector(rotsun)
mhor = astronomy.Horizon(besttime, observer, meq.ra, meq.dec, astronomy.Refraction.Airless)
shor = astronomy.Horizon(besttime, observer, seq.ra, seq.dec, astronomy.Refraction.Airless)
print(mhor.altitude, shor.altitude, mhor.altitude - shor.altitude)
# 7.162904852724097 -7.808661756113622 14.971566608837719
```

So now the discrepancy between Astronomy Engine and Swiss Ephemeris is 0.001163 degrees, or 4.2 arcseconds, which is about the amount of error I would expect.

I remembered there is an `EquatorFromVector` function, so we don't need the undocumented `_vector2radec` after all.

Also note that I changed to correcting for aberration in the GeoVector calls, because it makes a significant difference for the Sun. It makes no difference at all for the Moon, but I made them both `True` for consistency.

**Reply by hidp123** [reply]
Thanks for your efforts and help. I'll give this a go.

**Comment by hidp123** [comment]
I was wondering if this is similar to the work you did here https://github.com/cosinekitty/astronomy/issues/279, but instead of going higher we are going lower towards the core of the Earth. So a negative altitude of sorts, and the horizon will shrink equally as well, like a smaller sphere.

**Reply by cosinekitty** [reply]
The distinction between geocentric altitude and geodetic altitude is a different problem. This has to do with corrections for the oblateness of the Earth. If you are on a ship 45 degrees north of the equator, and you measure the altitude angle of an object above the sea level horizon, it turns out that your horizon plane is not perpendicular to a line passing through you and the geocenter. In other words, if you let a plumb line hang, and extrapolate the direction it points, that extended line will not intersect with the Earth's center. It will miss because the rotation of the Earth causes the ocean (and your sense of up and down) to bulge away from a spherical shape. Stated another way, your sense of gravitational up and down is a combination of a gravity vector and a centrifugal force vector that point in different directions.

**Reply by YingChangJ** [reply]
Hello, 

I happened to see that you are still active with this problem few days ago, and try to reproduce the result with the three libraries you mentioned, they could produce pretty much the same result (astronomy-engine 0.0001 diff, and skyfield 0.00001 diff with swisseph). 
As for astronomy-engine, starting with GeoVector not bad, we can get all others with rotation matrix. 
```
import astronomy

observer = astronomy.Observer(39, -76.8)
besttime = astronomy.Time.Make(1990, 11, 19, 22, 29, 11)
geomoon = astronomy.GeoVector(astronomy.Body.Moon, besttime, True)
rot1 = astronomy.Rotation_EQJ_ECL()
rot2 = astronomy.Rotation_ECL_HOR(besttime, observer)
rot3 = astronomy.CombineRotation(rot1, rot2)
mec_hor = astronomy.RotateVector(rot3, geomoon)
print(astronomy.SphereFromVector(mec_hor))

# Spherical(lat=7.162904852724086, lon=132.47079289201685, dist=0.002715735744607922)
```
It is exactly the same as the one above. (while swisseph give 7.163034615148587)
(I post it here to show how the other two libraries could match, in your recent post)
